File: backend/app/schemas/voice_note.py
from pydantic import BaseModel, Field


class UploadUrlRequest(BaseModel):
    duration_ms: int = Field(..., gt=0, le=60000)
    content_type: str = Field(..., pattern=r"^audio/(mp4|aac|mp3)$")


class UploadUrlResponse(BaseModel):
    upload_id: str
    upload_url: str
    expires_at: str


class VoiceNoteCreateRequest(BaseModel):
    upload_id: str = Field(
        ...,
        pattern=r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$",
    )
    duration_ms: int = Field(..., gt=0, le=60000)
    audio_size: int = Field(default=0, ge=0)
    post_to_farcaster: bool = True
    cast_text: str = Field(default="", max_length=1024)


class VoiceNoteAuthor(BaseModel):
    fid: int
    username: str
    display_name: str
    pfp_url: str | None = None


class VoiceNoteResponse(BaseModel):
    id: str
    fid: int
    duration_ms: int
    audio_url: str
    audio_size: int
    waveform_peaks: list[float] | None = None
    transcript: str | None = None
    transcript_lang: str | None = None
    transcript_words: list[dict] | None = None
    cast_hash: str | None = None
    cast_url: str | None = None
    caption: str | None = None
    created_at: str
    deleted_at: str | None = None


class VoiceNoteDetailResponse(BaseModel):
    voice_note: VoiceNoteResponse
    author: VoiceNoteAuthor
    reaction_counts: dict[str, int]
    viewer_reactions: list[str]
    play_count: int


class VoiceNoteFeedResponse(BaseModel):
    voice_notes: list[VoiceNoteDetailResponse]
    next_cursor: str | None = None


class PlayRequest(BaseModel):
    listened_ms: int = Field(..., ge=0)
    completed: bool = False


class ReactionRequest(BaseModel):
    reaction_type: str = Field(..., pattern=r"^(like|recast)$")

File: farcaster-audio/types/voiceNote.ts
export interface VoiceNoteAuthor {
  fid: number;
  username: string;
  display_name: string;
  pfp_url: string | null;
}

export interface VoiceNote {
  id: string;
  fid: number;
  duration_ms: number;
  audio_url: string;
  audio_size: number;
  waveform_peaks: number[] | null;
  transcript: string | null;
  transcript_lang: string | null;
  transcript_words: Array<{
    word: string;
    start_ms: number;
    end_ms: number;
  }> | null;
  cast_hash: string | null;
  cast_url: string | null;
  parent_cast_hash: string | null;
  caption: string | null;
  created_at: string;
  deleted_at: string | null;
}

export interface VoiceNoteDetail {
  voice_note: VoiceNote;
  author: VoiceNoteAuthor;
  reaction_counts: Record<string, number>;
  viewer_reactions: string[];
  play_count: number;
}

export interface VoiceNoteFeedResponse {
  voice_notes: VoiceNoteDetail[];
  next_cursor: string | null;
}

export interface UploadUrlResponse {
  upload_id: string;
  upload_url: string;
  expires_at: string;
}

/** Union type for mixed feed items (cast or voice note) */
export type FeedItem =
  | { type: "cast"; data: import("@/types/neynar").NeynarCast }
  | { type: "voice_note"; data: VoiceNoteDetail };
